### PR TITLE
For mozilla-mobile#24461 increase touch target for scan, search engine & voice search button

### PR DESCRIPTION
**Pull request by jknair0**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

**After updating Search engine tap area**
1_updated_search_engine

**After updating Scan tap area**
I have updated search engine tap area first. ASA i updated scan tap area it started showing border for search engine. i think thats a false positive(or possibly cause by colliding tap area rect of scan).

2_updated_scan

**After updating voice search tap area**
updated_micro_phone

The close button does not seems to be in this project. Its in browser-toolbar dependency. so could not access it.

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package org.mozilla.fenix.search.toolbar
+
+import android.view.View
+import mozilla.components.concept.toolbar.Toolbar
+import org.mozilla.fenix.ext.increaseTapArea
+
+class IncreaseDpsAction(
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt:7]: Nice approach 👍.
This seems to be a decorator, so could we name it as such to have the intent clearer? Something like `IncreasedTapAreaActionDecorator`?
Adding a small class documentation would also help, and a small test for the bind method would also be nice.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package org.mozilla.fenix.search.toolbar
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt:1]: Please add the Mozilla license at the top of the file
```
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    override fun bind(view: View) {
+        action.bind(view)
+        view.increaseTapArea(TAP_INCREASE_DPS)
+    }
+
+    companion object {
+        private const val TAP_INCREASE_DPS = 8
+    }
+
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt:19]: nit: unneeded line.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+}
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt:20]: nit: no newline. (will be reported in static analysis)

File: app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt [hunk 5]
```diff
@@ -252,8 +254,11 @@ class SearchDialogFragment : AppCompatDialogFragment(), UserInteractionHandler {
             else -> {}
         }
 
-        binding.searchEnginesShortcutButton.setOnClickListener {
-            interactor.onSearchShortcutsButtonClicked()
+        binding.searchEnginesShortcutButton.apply {
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt:257]: nit: just for consistency, for `qrScanButton` below we have the `setOnClickListener` and `increaseTapArea` on separate lines. Can we have the same here?

**Review by Mugurell — COMMENTED** [review]
This looks great, thank you!
Seems almost ready to land, just some minor style changes.

Regarding 
> ASA i updated scan tap area it started showing border for search engine. i think thats a false positive(or possibly cause by colliding tap area rect of scan)

This happens because of how our current `increaseTapArea` method works.
It [sets a new TouchDelegate](https://github.com/mozilla-mobile/fenix/blob/54e17f4e2d926616f0cc3e6fd42371916afd9a4a/app/src/main/java/org/mozilla/fenix/ext/View.kt#L29) everytime it's called. So only the last one that is set will produce the needed effects.
For our scenario looks like we'll need one `TouchDelegate` delegating touches to multiple others. The solution does become more complex so we can also merge this now and leave that bit for later.

Agree with
> The close button does not seems to be in this project. Its in browser-toolbar dependency. so could not access it.

This will need deeper investigation especially based on the scenario described above.

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt (hunk 1 above)
Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt]: updated!

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt (hunk 2 above)
Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt]: updated

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt (hunk 3 above)
Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt]: fixed and ran `gradlew detekt` and `gradlew ktlint`. It succeeded. Please do help with command if it is being verified differently

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt (hunk 4 above)
Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt]: done

File: app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt (hunk 5 above)
Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt]: fixed

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt (hunk 3 above)
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt]: Those are indeed the commands to be used. Thank you!

File: app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt (hunk 1 above)
Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/IncreaseDpsAction.kt]: added the test as well!

**Review by Mugurell — APPROVED** [review]
This looks great! Thank you!

**Comment by Mugurell** [comment]
Opened https://github.com/mozilla-mobile/fenix/pull/24602 to run the complete suite of tests and land the patch.
Closing this.

File: app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt [hunk 6]
```diff
@@ -636,14 +640,16 @@ class SearchDialogFragment : AppCompatDialogFragment(), UserInteractionHandler {
...
-                )
+            val voiceSearchAction = BrowserToolbar.Button(
+                AppCompatResources.getDrawable(requireContext(), R.drawable.ic_microphone)!!,
+                requireContext().getString(R.string.voice_search_content_description),
+                visible = { true },
+                listener = ::launchVoiceSearch
             )
+            toolbarView.view.run {
+                addEditActionEnd(IncreasedTapAreaActionDecorator(voiceSearchAction))
+                invalidateActions()
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt:651]: @jknair0 I am wondering if `invalidateActions()` was necessary here?

Comment (jknair0) [reply app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt:651]: I have noticed that the bind of the actions was not getting invoked. So I had to do it manually.

**Pull request closed without merging**